clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java

```diff
@@ -344,72 +346,124 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
     @Test
     public void testExpiredBatches() throws InterruptedException {
         long retryBackoffMs = 100L;
-        long lingerMs = 3000L;
-        int requestTimeout = 60;
-
-        RecordAccumulator accum = new RecordAccumulator(1024, 10 * 1024, CompressionType.NONE, lingerMs, retryBackoffMs, metrics, time);
-        int appends = 1024 / msgSize;
+        long lingerMs = 500L;
+        int requestTimeoutMs = 2000;
+        int batchSize = 1024;
+        long totalSize = 10 * 1024;
+        long metadataMaxAgeMs = 5 * 60 * 1000L; // 5 min
+        int retries = 1;
+        long staleMetadataAgeMs  = Sender.getMetadataStaleMs(metadataMaxAgeMs, requestTimeoutMs, retryBackoffMs, retries);
+        List<RecordBatch> expiredBatches;
+        RecordAccumulator.ReadyCheckResult result;
+        Set<Node> readyNodes;
+        Map<Integer, List<RecordBatch>> drained;
+        boolean isMetadataStale = false;
+
+        assertTrue("Stale metadata age must be more than request timeout", staleMetadataAgeMs > metadataMaxAgeMs);
+        
+        Metadata metadata = new Metadata(retryBackoffMs, metadataMaxAgeMs, false, new ClusterResourceListeners());
+        metadata.update(cluster, time.milliseconds());
+        RecordAccumulator accum = new RecordAccumulator(batchSize, totalSize, CompressionType.NONE, lingerMs, retryBackoffMs, metrics, time);
+        int appends = batchSize / msgSize;
 
         // Test batches not in retry
         for (int i = 0; i < appends; i++) {
             accum.append(tp1, 0L, key, value, null, maxBlockTimeMs);
             assertEquals("No partitions should be ready.", 0, accum.ready(cluster, time.milliseconds()).readyNodes.size());
         }
-        // Make the batches ready due to batch full
+        // subtest1: Make the batches ready due to batch full
```

Comment:
Rather than having one giant test which includes these subtests, could we split them into separate test cases? It's easier to debug failures that way since the test case is more isolated. It also lets you provide a meaningful name to the test which makes the nature of the failure clearer just looking at the output from the build.